Implement custom prometheus http handler

alexg-axis opened the pull request:
Implement a custom prometheus http handler in order to:

1. Support listening on a specified address as opposed to any
2. Remove the requirement on the unmaintained promhttp library

This feature comes with one limitation: if an IPv4 address is used, the server will not listen on the IPv6-mapped address, even if IPv6 is available. That is, dual-stacking does not work.

Solves: #1475

eakraly commented:
Thanks @alexg-axis  - this is an awesome contribution!
Getting rid of libpromhttp gets us half way to remove unsupported libraries from DO

eakraly commented on src/apps/relay/prom_server.c:
```diff
@@ -127,7 +161,36 @@ void start_prometheus_server(void) {
+
+  ioa_addr server_addr;
+  addr_set_any(&server_addr);
+  if (turn_params.prometheus_address[0]) {
```
What about default? Maybe having it open to localhost?

eakraly commented on src/apps/relay/prom_server.c:
```diff
@@ -1,5 +1,6 @@
 #include "prom_server.h"
 #include "mainrelay.h"
+#include "microhttpd.h"
```
Already included in .h file - why another include?

alexg-axis replied:
Good spot. Removed.

eakraly commented:
Also, could you please add some test that validates prometheus interface is up and running?  See examples folder where you can add another test
Also, you missed a reference to promhttp in .github folder

eakraly commented on src/apps/relay/prom_server.c:
```diff
@@ -35,6 +35,44 @@ prom_counter_t *turn_total_traffic_peer_sentb;
 
 prom_gauge_t *turn_total_allocations;
 
+#if MHD_VERSION >= 0x00097500
```
Seems like MHD_Result was introduced in libmicrohttpd 4 years ago in 0.9.71, not 0.9.75
https://github.com/Karlson2k/libmicrohttpd/commit/6347f514aa2388e774d5bf356df8046864e5f73c

Also, this was not in promhttp - so we are not losing anything by not having `#if MHD_VERSION >= 0x00097500`

I suggest to simplify the code and require libmicrohttpd >= 0.9.71
What do you think?

alexg-axis replied:
Yeah, I think it makes a lot of sense. There's some other code also checking the version of microhttpd, like the error log and epoll flags. Should we remove those ifdefs in this PR or take that another time?

eakraly replied:
Please remove it now - no point in adding and then removing code.
Thanks!

alexg-axis replied:
The CI is installing libmicrohttpd 0.9.66-1, seemingly due to ubuntu:20.04 being used. If I try to run ubuntu:22.04 instead, I get 0.9.75. 

I think I've seen a few issues and PRs on this topic. I guess the "easiest" thing to do is to update the CI/CD and start requiring newer versions of ubuntu. Otherwise we should just be able to revert my latest commit, again adding support for <0.9.71.

eakraly replied:
Ugh. Let's use `int` return type instead of `MHD_Status` - ugly but should work I hope.

eakraly replied:
I see. Let's do that (revert). Maybe you can make it a little nicer by defining `MHD_Result` as int for older version and keeping the code in its new form
```
#if MHD_VERSION < 0x00097500
define MHD_Result int
#endif

MHD_Result promhttp_handler(....
```

alexg-axis replied:
Note that it's `enum MHD_Result`. I defined a `MHD_RESULT` that works the way you described. Does that work for you?

Also, I stumbled upon a changelog entry for the change. It recommends using `MHD_VERSION >= 0x00097002`. So I changed that as well. Here's a link to the changelog: https://github.com/Karlson2k/libmicrohttpd/blob/a1085032dc631e387eb8e06b41a382e7dc6dd8eb/ChangeLog#L1055.

alexg-axis commented:
@eakraly I noticed a potentially problematic behavior I can't really explain. I don't think it's related to these changes, but it's worth bringing up.

Every now and then the server seems to return empty metrics.

For example:

```
curl --silent localhost:9641/metrics | grep -A1 turn_total_traffic_rcvb
# HELP turn_total_traffic_rcvb Represents total finished sessions received bytes
# TYPE turn_total_traffic_rcvb counter
turn_total_traffic_rcvb 1484
```

```
curl --silent localhost:9641/metrics | grep -A1 turn_total_traffic_rcvb
# HELP turn_total_traffic_rcvb Represents total finished sessions received bytes
# TYPE turn_total_traffic_rcvb counter
turn_total_traffic_rcvb 1484
```

```
curl --silent localhost:9641/metrics | grep -A1 turn_total_traffic_rcvb
# HELP turn_total_traffic_rcvb Represents total finished sessions received bytes
# TYPE turn_total_traffic_rcvb counter
```

```
curl --silent localhost:9641/metrics | grep -A1 turn_total_traffic_rcvb
# HELP turn_total_traffic_rcvb Represents total finished sessions received bytes
# TYPE turn_total_traffic_rcvb counter
turn_total_traffic_rcvb 1484
```

When the prom lib is generating the response body, there are a few early returns that I guess could result in metrics missing. The error code is unhandled by `prom_collector_registry_bridge`.

https://github.com/digitalocean/prometheus-client-c/blob/c57034d196582d99267d027abb52a05a55dc07f6/prom/src/prom_collector_registry.c#L215

In grafana, for example, I guess it's quite common to fill in "null" values by repeating the last value. But perhaps this is a symptom of something else?

eakraly commented:
> @eakraly I noticed a potentially problematic behavior I can't really explain. I don't think it's related to these changes, but it's worth bringing up. I've only noticed it once, so it might just be due to something unique to my environment.
> 
> Every now and then the server seems to return empty metrics.

Interesting. Do you observe it without your change? I bet it is unrelated. If it is - we can go on and create another issue to tackle this specific problem.

alexg-axis commented:
@eakraly I've now built and tested the code manually with asan and things work as expected without any issues. I have also added an option to control the base path (or the full path). This helps the prometheus server to run behind a reverse proxy on a specified path as coturn can be told to use that path.

In https://github.com/coturn/coturn/pull/1591/commits/e5cec04403539ce513b0ed37bf8e0724480fdedb I've opted for specifying a base path and "hard coding" the `/metrics` path. But as that is the only path today, I also wrote https://github.com/coturn/coturn/pull/1591/commits/1535c9605b473a2ad429b5085c4e4531704ce89d to allow users to specify the full path instead.

What are your thoughts on that option? I want to be able to expose prometheus behind a reverse proxy on a specified path and it would be nice if that path could for example be `/metrics/coturn` without coturn requiring `/metrics` to be added to the end, ending up with the repetitive `/metrics/coturn/metrics` path.

One potential issue I see is that it might be more cumbersome to add additional endpoints to the server handler (whatever they may be) as now `/metrics` is a common base path for all endpoints, so adding an endpoint would result in `/metrics/xyz`, where you might have wanted `/metrics` and `/xyz`.

eakraly commented:
Thanks @alexg-axis !

1) Having default to be equal to today's configuration is great - that means no breaking change.
2) I've seen a lot of software that exposes metrics on `/metrics` because that's the prometheus default and when you scrape multiple targets - they have to expose on different ports anyway. Also, it means that prometheus need to scrape multiple targets (let's say coturn and node_exporter) anyway - whether they expose it on `/metric/coturn` or `/metrics`.

> What are your thoughts on that option?

I'm ok with the change - IFF that solves a problem for you (though I do not really see what it would be - if you can expand on your reverse proxy config and how they will work together better it would be great)

> One potential issue I see is ....

I do not understand your concern - the handler you edited is prom specific HTTP handler. In any case, this is a simple if..else flow that can match more paths in the future (let's say we split prom into 2 subsets of metrics /metrics/a and /metrics/b - we can still match them).

eakraly commented:
@alexg-axis my additional suggestion would be to split this PR into
1) Custom prom http handler (as it was before last 2 commits)
2) Custom prom path setting (last 2 commits)

These are 2 different things and better have them separately (just in case there is a fallout and we need to bisect the regression)

eakraly commented:
> > @alexg-axis my additional suggestion would be to split this PR into
> > 
> > 1. Custom prom http handler (as it was before last 2 commits)
> > 2. Custom prom path setting (last 2 commits)
> > 
> > These are 2 different things and better have them separately (just in case there is a fallout and we need to bisect the regression)
> 
> I've rebased this PR and removed the addition of base path handling. Let's take that discussion separately as you said ( #1602. I'll rebase once this is in). Regarding testing, could you clarify what you would want to add before we can get this in?
> 
> Would you like me to change `examples/run_tests.sh` to start turnserver with prometheus enabled on localhost for example? Or would you like an additional `examples/run_tests_prom.sh` that starts prometheus with the different ways to configure prometheus?

For testing I think separate `examples/run_tests_prom.sh` would be better - prom is not built for all and we may see a bunch of failures changing existing test setup. Creating new one gives us opportunity to expand for more prom tests in the future.

alexg-axis commented:
> > > @alexg-axis my additional suggestion would be to split this PR into
> > > 
> > > 1. Custom prom http handler (as it was before last 2 commits)
> > > 2. Custom prom path setting (last 2 commits)
> > > 
> > > These are 2 different things and better have them separately (just in case there is a fallout and we need to bisect the regression)
> > 
> > 
> > I've rebased this PR and removed the addition of base path handling. Let's take that discussion separately as you said ( #1602. I'll rebase once this is in). Regarding testing, could you clarify what you would want to add before we can get this in?
> > Would you like me to change `examples/run_tests.sh` to start turnserver with prometheus enabled on localhost for example? Or would you like an additional `examples/run_tests_prom.sh` that starts prometheus with the different ways to configure prometheus?
> 
> For testing I think separate `examples/run_tests_prom.sh` would be better - prom is not built for all and we may see a bunch of failures changing existing test setup. Creating new one gives us opportunity to expand for more prom tests in the future.

I've added a few tests. Could you trigger the CI pipeline so we can see them work there as well?

eakraly commented on examples/run_tests_prom.sh:
```diff
@@ -0,0 +1,59 @@
+
+function assert_prom_response() {
+  # Match something that looks like the expected body
+  curl --silent --fail "$1" | grep 'TYPE\|HELP\|counter\|gauge' >/dev/null
```
Please use wget instead of curl - it is not installed

alexg-axis replied:
@eakraly is there anything else needed from our side?

eakraly commented on examples/run_tests_prom.sh:
```diff
@@ -0,0 +1,59 @@
+}
+
+echo "Running without prometheus"
+../bin/turnserver &
```
> /dev/null

eakraly replied:
> /dev/null

eakraly commented on examples/run_tests_prom.sh:
```diff
@@ -0,0 +1,59 @@
+kill "$turnserver_pid"
+
+echo "Running turnserver with prometheus, using defaults"
+../bin/turnserver --prometheus &
```
> /dev/null

eakraly commented on examples/run_tests_prom.sh:
```diff
@@ -0,0 +1,59 @@
+kill "$turnserver_pid"
+
+echo "Running turnserver with prometheus, using custom address"
+../bin/turnserver --prometheus --prometheus-address="127.0.0.1" &
```
> /dev/null

eakraly commented on examples/run_tests_prom.sh:
```diff
@@ -0,0 +1,59 @@
+kill "$turnserver_pid"
+
+echo "Running turnserver with prometheus, using custom port"
+../bin/turnserver --prometheus --prometheus-port="8080" &
```
> /dev/null

alexg-axis commented:
@eakraly thanks for merging! I've updated https://github.com/coturn/coturn/pull/1602 to add a flag to specify at which path the metrics should be served.
